src/filter/filter_gain.rs

```diff
@@ -0,0 +1,84 @@
+use nalgebra as na;
+use super::super::config::*;
+
+// pub fn measurement_vector(
+//     jacobian: &Mat5,
+//     previous_state_vec: &Vec5) -> Vec2{
+//     return jacobian * previous_state_vec
+// }
+
+pub fn state_vector( // x
+    extrap_state_vector: &Vec5, 
+    kalman_gain: &Mat5x2, 
+    measurement : &Vec2, 
+    sensor_mapping: &Mat2x5) -> Vec5 {
+
+    let parens = measurement - (sensor_mapping * extrap_state_vector);
+    let kalman_product = kalman_gain * parens;
+    return extrap_state_vector + kalman_product;
+}
+
+//TODO: remove `unwrap` on the inverse
+#[allow(dead_code)]
+pub fn kalman_gain ( // K
+    C : &Mat5,
+    H : &Mat2x5,
+    V : &Mat2) -> Mat5x2 {
+
+    let parens = V + ( H * C * H.transpose() );
+    let kalman_gain = C * H.transpose() * parens.try_inverse().unwrap();
+    
+    kalman_gain
+}
+
+//TODO add lazy static for identity
```

Comment:
I dont see anything mentioning `nalgebra`'s `MatrixMN` being threadsafe in the [docs](https://docs.rs/nalgebra/0.18.0/nalgebra/base/type.MatrixMN.html), but seeing as this compiles I think we are in good territory:

```
use config::*;
fn test_threadsafe<T: Send + Sync>(input: &T) {}

fn construct_identity() {
    lazy_static!{
        static ref identity: Mat4 = Mat4::identity();
    }
    test_threadsafe(&identity);
}
```

As for compile time optimization, I had unexpected runtime errors with `nagebra` at some points, so I was hesitant to trust that this would be compiled away. This comment is more of a reminder to make sure that `identity` becomes static when compiled.

Reply:
Ok, sure. `nalgebra`'s matrices are not thread-safe in all likelihood. I was just wondering if a lazily initialized static would be thread safe in general, not specifically the matrix identity here.

Reply:
Yup, lazy static works more or less like static function var initialization in C++ and java : initialize on first access, block other threads during the initialization process. Unlike in those languages, however, there is a requirement of thread-safe accesses (e.g. only reads) after initialization.

Reply:
👍